File: Engines/BaselineEngine.swift
import Foundation

struct BaselineResult {
    let mean: Double
    let standardDeviation: Double
    let sampleCount: Int
    let windowDays: Int
}

struct BaselineEngine {
    /// Compute a 28-day rolling baseline for a set of daily values.
    /// Falls back to shorter windows (minimum 3 samples).
    static func computeBaseline(
        values: [Double],
        windowDays: Int = HealthKitConstants.baselineWindowDays,
        minimumSamples: Int = HealthKitConstants.minimumBaselineSamples
    ) -> BaselineResult? {
        guard !values.isEmpty else { return nil }

        // Use up to windowDays most recent values
        let recentValues = Array(values.suffix(windowDays))

        guard recentValues.count >= minimumSamples else {
            // Not enough data; try with what we have if at least minimum
            if values.count >= minimumSamples {
                let fallbackValues = Array(values.suffix(minimumSamples))
                return BaselineResult(
                    mean: fallbackValues.map { $0 }.mean,
                    standardDeviation: fallbackValues.map { $0 }.standardDeviation,
                    sampleCount: fallbackValues.count,
                    windowDays: fallbackValues.count
                )
            }
            return nil
        }

        let mean = recentValues.mean
        let stddev = recentValues.standardDeviation

        return BaselineResult(
            mean: mean,
            standardDeviation: max(stddev, 0.001), // Prevent division by zero
            sampleCount: recentValues.count,
            windowDays: windowDays
        )
    }

    /// Compute z-score for a value against a baseline.
    static func zScore(value: Double, baseline: BaselineResult) -> Double {
        guard baseline.standardDeviation > 0 else { return 0 }
        return (value - baseline.mean) / baseline.standardDeviation
    }

    /// Update a running baseline with a new value.
    /// Returns the updated baseline using exponential weighting.
    static func updateBaseline(
        current: BaselineResult,
        newValue: Double,
        alpha: Double = 0.1
    ) -> BaselineResult {
        let newMean = current.mean * (1 - alpha) + newValue * alpha
        let newVariance = pow(current.standardDeviation, 2) * (1 - alpha) + pow(newValue - newMean, 2) * alpha
        let newStdDev = sqrt(max(newVariance, 0.001))

        return BaselineResult(
            mean: newMean,
            standardDeviation: newStdDev,
            sampleCount: current.sampleCount + 1,
            windowDays: current.windowDays
        )
    }
}

File: android/core/engine/src/main/kotlin/com/apexfit/core/engine/BaselineEngine.kt
package com.apexfit.core.engine

import kotlin.math.max
import kotlin.math.pow
import kotlin.math.sqrt

data class BaselineResult(
    val mean: Double,
    val standardDeviation: Double,
    val sampleCount: Int,
    val windowDays: Int,
) {
    val isValid: Boolean
        get() = sampleCount >= 3 && standardDeviation > 0
}

object BaselineEngine {

    fun computeBaseline(
        values: List<Double>,
        windowDays: Int = 28,
        minimumSamples: Int = 3,
    ): BaselineResult? {
        if (values.isEmpty()) return null

        val recentValues = values.takeLast(windowDays)

        if (recentValues.size >= minimumSamples) {
            return BaselineResult(
                mean = recentValues.mean,
                standardDeviation = max(recentValues.standardDeviation, 0.001),
                sampleCount = recentValues.size,
                windowDays = windowDays,
            )
        }

        // Fallback: use all available data if enough samples
        if (values.size >= minimumSamples) {
            val fallbackValues = values.takeLast(minimumSamples)
            return BaselineResult(
                mean = fallbackValues.mean,
                standardDeviation = max(fallbackValues.standardDeviation, 0.001),
                sampleCount = fallbackValues.size,
                windowDays = fallbackValues.size,
            )
        }

        return null
    }

    fun zScore(value: Double, baseline: BaselineResult): Double {
        if (baseline.standardDeviation <= 0) return 0.0
        return (value - baseline.mean) / baseline.standardDeviation
    }

    fun updateBaseline(
        current: BaselineResult,
        newValue: Double,
        alpha: Double = 0.1,
    ): BaselineResult {
        val newMean = current.mean * (1 - alpha) + newValue * alpha
        val newVariance = current.standardDeviation.pow(2) * (1 - alpha) +
            (newValue - newMean).pow(2) * alpha
        val newStdDev = sqrt(max(newVariance, 0.001))

        return BaselineResult(
            mean = newMean,
            standardDeviation = newStdDev,
            sampleCount = current.sampleCount + 1,
            windowDays = current.windowDays,
        )
    }
}
